algebra-core/src/curves/models/bn/mod.rs

```diff
@@ -0,0 +1,219 @@
+        let mut c1 = coeffs.1;
+        let mut c2 = coeffs.2;
+
+        match P::TWIST_TYPE {
+            TwistType::M => {
+                c2.mul_assign_by_fp(&p.y);
+                c1.mul_assign_by_fp(&p.x);
+                f.mul_by_014(&c0, &c1, &c2);
+            }
+            TwistType::D => {
+                c0.mul_assign_by_fp(&p.y);
+                c1.mul_assign_by_fp(&p.x);
+                f.mul_by_034(&c0, &c1, &c2);
+            }
+        }
+    }
+
+    fn exp_by_neg_x(mut f: Fp12<P::Fp12Params>) -> Fp12<P::Fp12Params> {
+        f = f.cyclotomic_exp(&P::X);
+        if !!!P::X_IS_NEGATIVE {
+            f.conjugate();
+        }
+        f
+    }
+}
+
+impl<P: BnParameters> PairingEngine for Bn<P> {
+    type Fr = <P::G1Parameters as ModelParameters>::ScalarField;
+    type G1Projective = G1Projective<P>;
+    type G1Affine = G1Affine<P>;
+    type G1Prepared = G1Prepared<P>;
+    type G2Projective = G2Projective<P>;
+    type G2Affine = G2Affine<P>;
+    type G2Prepared = G2Prepared<P>;
+    type Fq = P::Fp;
+    type Fqe = Fp2<P::Fp2Params>;
+    type Fqk = Fp12<P::Fp12Params>;
+
+    fn miller_loop<'a, I>(i: I) -> Self::Fqk
+    where
+        I: IntoIterator<Item = &'a (Self::G1Prepared, Self::G2Prepared)>,
+    {
+        let mut pairs = vec![];
+        for (p, q) in i {
+            if !p.is_zero() && !q.is_zero() {
+                pairs.push((p, q.ell_coeffs.iter()));
+            }
+        }
+
+        let mut f = Self::Fqk::one();
+
+        for i in (1..P::ATE_LOOP_COUNT.len()).rev() {
+            if i != P::ATE_LOOP_COUNT.len() - 1 {
+                f.square_in_place();
+            }
+
+            for (p, ref mut coeffs) in &mut pairs {
+                Self::ell(&mut f, coeffs.next().unwrap(), &p.0);
+            }
+
+            let bit = P::ATE_LOOP_COUNT[i - 1];
+            match bit {
+                1 => {
+                    for &mut (p, ref mut coeffs) in &mut pairs {
+                        Self::ell(&mut f, coeffs.next().unwrap(), &p.0);
+                    }
+                }
+                -1 => {
+                    for &mut (p, ref mut coeffs) in &mut pairs {
+                        Self::ell(&mut f, coeffs.next().unwrap(), &p.0);
+                    }
+                }
+                _ => continue,
+            }
+        }
+
+        if P::ATE_LOOP_COUNT_IS_NEGATIVE {
+            f.conjugate();
+        }
+
+        for &mut (p, ref mut coeffs) in &mut pairs {
+            Self::ell(&mut f, coeffs.next().unwrap(), &p.0);
+        }
+
+        for &mut (p, ref mut coeffs) in &mut pairs {
+            Self::ell(&mut f, coeffs.next().unwrap(), &p.0);
+        }
+
+        f
+    }
+
+    fn final_exponentiation(f: &Self::Fqk) -> Option<Self::Fqk> {
+        /*
+          Easy part: result = elt^((q^6-1)*(q^2+1)).
+          Follows, e.g., Beuchat et al page 9, by computing result as follows:
+             elt^((q^6-1)*(q^2+1)) = (conj(elt) * elt^(-1))^(q^2+1)
+        */
```

Comment:
Just for consistency with the rest of the library

```suggestion
        // Easy part: result = elt^((q^6-1)*(q^2+1)).
        // Follows, e.g., Beuchat et al page 9, by computing result as follows:
        //   elt^((q^6-1)*(q^2+1)) = (conj(elt) * elt^(-1))^(q^2+1)
```